File: Problems/ cs50-problems-2021-x-sentimental-credit/credit.py
from cs50 import get_int


def main():
    card = get_int("Enter card number: ")
    length = len(str(card))
    flag = 0
    visa = 0
    mc = 0
    ax = 0
    if length < 13 or length > 16:
        flag = 1
        
    else:
        
        if length == 13 or length == 14:
            
            temp = card
            while len(str(temp)) != 1:
                temp = int(temp / 10)
                
            if temp == 4:
                returnType = counting(card)
                
                if returnType == 1:
                    visa = 1
                else: 
                    flag = 1
                
            else: 
                flag = 1
            
        elif length == 15:
            
            temp = card
            while len(str(temp)) != 2:
                temp = int(temp / 10)
                
            firstTemp = int(temp / 10)
            if temp == 34 or temp == 37 or firstTemp == 4:
                returnType = counting(card)
                if returnType == 1:
                    if firstTemp == 4:
                        visa = 1
                    else:
                        ax = 1
                else: 
                    flag = 1
            else: 
                flag = 1
            
        else:
            
            temp = card
            while len(str(temp)) != 2:
                temp = int(temp / 10)
                
            firstTemp = int(temp / 10)
            
            if temp == 51 or temp == 52 or temp == 53 or temp == 54 or temp == 55 or firstTemp == 4:
                returnType = counting(card)
                if returnType == 1:
                    if firstTemp == 4:
                        visa = 1
                    else:
                        mc = 1
                else:
                    flag = 1
              
            else:
                flag = 1
    if flag == 1:
        print("INVALID")
    elif visa == 1:
        print("VISA")
    elif ax == 1:
        print("AMEX")
    else:
        print("MASTERCARD")


def counting(cardNum):
    cnt = 1
    total = 0
    
    while cardNum != 0:
        n = cardNum % 10
        
        if cnt % 2 == 0:
            n *= 2
            
        while n != 0:
            total += (n % 10)
            n = int(n/10)
            
        cnt += 1
        cardNum = int(cardNum / 10)    

    if total % 10 == 0:
        return 1
    else:
        return 0
    
    
main()

File: Problems/ cs50-problems-2021-x-credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void){

    //asking user for number
    
    long long number=get_long("Number: \n");
    
    //variables
    
    int cnt = 1, type = 0;
    long long total_sum, sum_product = 0, sum_digits = 0;
    
    //getting each digit in the number by dividing by 10 until it reaches 0
    
    while(number != 0){
        
        // checking if the number consists of 2 digits only to detect card type
        
        if(number / 100 == 0 && type == 0)
           type = number;

        // if cnt: odd dont multiply but sum else multiply
        
        if(cnt % 2 != 0){ 
            
            // add to the sum the last digit
            
            sum_digits += (number % 10);
            
        }
        else {
            //assign last digit in a variable
            
            int digit = (number % 10);
            
            // multiply digit by 2 and loop through it to add its digits
            
            digit *= 2;
            
            while(digit != 0){
                sum_product += (digit % 10);
                digit /= 10;
            }
            
        }
        
        // constantly removing last digit of the number
        
        number /= 10; 
        cnt++;
    }
    
    total_sum = sum_product + sum_digits;
     
     //if it ends with 0 valid else invalid
     
     if(total_sum % 10 == 0 && cnt > 12) { 
         
         if( type % 100 == 34 || type % 100 == 37)
           printf("AMEX\n");
           
        else if( type % 100 == 51 || type % 100 == 52 || type % 100 == 53 || type % 100 == 54 || type %100 == 55)
                    printf("MASTERCARD\n");
        else {
            type /= 10;
            if(type % 10 == 4)
                       printf("VISA\n");
            else printf("INVALID\n");

        }

     }
     else printf("INVALID\n");
    
}